### DISCUSSION TITLE
Replacing a (mostly) failed ZST39 Stick - Guidance/Opinions Wanted

### DISCUSSION THREAD

**Discussion by Botched1** [zwave-js/zwave-js#6875] [Q&A] [answered]
Per the thread below: https://github.com/zwave-js/node-zwave-js/discussions/6874

I flashed my ZST39 from 7.18.3 (1.1 in their versioning) to 7.21.3 (1.3 in their versioning). It did not go well, and I am having a lot of issues. Unfortunately in the flashing process my stick seems to be somewhat messed up. I cannot downgrade or install any gbl firmware to the ZST39 in Zwave JS UI or PC Controller. Both just error out.

The current stick is recognized in both ZWave JS UI and PC Controller. I can do an NVM backups in both ZWave JS UI and PC Controller.

I have a new ZST39 on the way. When it gets here I am hoping to be able to move everything to the new stick, and not lose/have to re-pair everything.

Question - Does anyone have any guidance on how to do this "the right way", and as importantly what NOT to do in that process?
* Is it as simple as restoring the NVM backup to the new stick since they are identical models?
* Should I use ZWave JS UI or PC Controller for the NVM restore? Does it matter (aka is one more reliable than the other for this)?

New stick gets here tomorrow, so I'm just trying to think through things today. Anything I can do to avoid having to re-pair everything would be wonderful.

Thanks!

**Comment by AlCalzone** [comment, marked as answer]
> Is it as simple as restoring the NVM backup to the new stick since they are identical models?

Assuming the NVM (backup) is not corrupt, then yes.

You'll want to use Z-Wave JS for restoring since it converts the NVM to the target format if you are moving to a lower firmware version for example.
PC controller just restores the backup 1:1

**Reply by Botched1** [reply]
> Assuming the NVM (backup) is not corrupt, then yes.

That will be the trick, and there is no way (I know of) to know if it is/is not until I try it... I will say the NVM backup is a reasonable size (40K or so - which is what they have typically been) so that is encouraging.

Worst comes to worst, I guess I'm in for 3 fun filled days of re-pairing everything.

I did try using Zwave JS UI to restore a ZST39 NVM backup to a UZB-7 I have... It failed, but I didn't put my time/thought into whether that should even work or not.

**Reply by Botched1** [reply]
Seems like there are no real "great" options in general.

I have 40-ish Z-Wave Long Range devices so want LR support. That keeps me in 700 or 800 controller territory.

* I have 2x UZB7 sticks, but I keep seeing that 700 controllers in general have a lot of issues so thought I should pass on that.
* I also have a z-wave.me Z-Station. 
  * I bought it because I like the form factor and antenna design. Well, it uses proprietary firmware and it seems that it can't be backed up/restored from Zwave JS (posted on their forum to ask if this is even intended to work)... I tried, it failed. Support from @poltos is not great in terms of getting it to work perfectly with Zwave JS (versus his z-way software), so I gave up and that went in the drawer for now. I didn't send it back to Amazon as I would like to support him (I also have a UZB and a razzberry 7 pro I don't use because of various issues).

So I thought I would go with an 800 based controller. 

* There is only one 800 series one I can readily buy - Zooz ZST39... 
 
So there are really not a lot of options out there if one wants LR support, and full Zwave JS compatibility...

**Reply by AlCalzone** [reply]
There is also an 800 series stick from HomeSeer now.

**Reply by Botched1** [reply]
Nice, I didn't know about that one (as the Z-Wave Alliance hardly ever updates the compatibility website any more).

Thanks!

**Reply by PoltoS** [reply]
@Botched1 I can't tell you about how Z-Wave JS is doing it, @AlCalzone is the right person. But I can say that copying the full NVM is a bad approach for backup & restore as the SDK can store data in different places from version to version. It is like copying the full HDD from Mac to Windows in hope that your personal files will be copied. They won't as different filesystems are used in both OS. It works between some dongles with some magic, but we can't guarantee it (I think, nobody can, it is not specified; NVM backup&restore is made to restore to the very same dongle on the very same SDK version, the rest might not work and if it does, it is just a luck).

Z-Wave.Me dongles (RaZberry 7, Z-Station and UZB) have own backup & restore method what works across 3/5/7/8th series. This one is 100% compatible despite the SDK version and filesystem structure on the dongle.

If you are using another dongle, we are releasing next week a new tool that will help you with the migration. It can port from ANY Z-Wave controller to Z-Station or RaZberry 7 (without using Z-Way, will work with Z-Wave JS too). If you would like, I'll invite you for testing it. Should I?

**Reply by Botched1** [reply]
> Z-Wave.Me dongles (RaZberry 7, Z-Station and UZB) have own backup & restore method what works across 3/5/7/8th series. This one is 100% compatible despite the SDK version and filesystem structure on the dongle.

It would be nice if that proprietary method of NVM backup/restore could be done in Z-Wave JS for those of use that want to use the z-wave.me controller there. Otherwise there is no way to backup/restore NVMs at all (without removing the device, putting in a different computer, using the z-wave.me tools).



> If you are using another dongle, we are releasing next week a new tool that will help you with the migration. It can port from ANY Z-Wave controller to Z-Station or RaZberry 7 (without using Z-Way, will work with Z-Wave JS too). If you would like, I'll invite you for testing it. Should I?

@PoltoS Yes, I would be interested in that. Even if I don't test it for my main production system, I have many different z-wave controllers, so can test a number of different combinations. 

In fact, if there were a beta available, I would be willing to test it right now since my current setup is broken, so I have nothing to lose! I have a Z-Station to test it on. I also have a Razzberry 7 Pro, but would have to rebuild the OS to test it with that.

I also just installed the updated z-way add-on in home assistant and am going to look at that again. That is off-topic for this forum, though, so can continue comments on that in your forum.

**Reply by PoltoS** [reply]
We can definitely share our way to do backup & restore. Nobody asked for it, so we never shared it.

**Reply by Botched1** [reply]
Any possibility of testing the conversion tool today or tomorrow, since I need to do something with my currently broken network anyway? :) If not, no worries.

**Reply by PoltoS** [reply]
Maybe on Friday. Let me check with the R&D tomorrow. It is currently in work.

**Reply by AlCalzone** [reply]
@PoltoS I'd also be interested in some details. Do you have a custom memory format across generations or do you implement target-specific conversions [like we do](https://github.com/zwave-js/node-zwave-js/tree/master/packages/nvmedit)? Our approach supports everything from SDK 6.61 to the current one, as long as there are no manufacturer-specific modifications to the NVM.

**Reply by PoltoS** [reply]
@AlCalzone none of them ;) we have custom restore function since 3rd gen sticks. This function says "save me the device #n with this generic/specific type. Same for home ID and node ID and routing table. The stick will save it correctly as it knows exactly how I should be saved for it's version.

It is implemented via NVM write operations, but internally the stick will store it on its own. I ca. Share with you the code we use. It is about 10-20 lines if code.

**Reply by PoltoS** [reply]
The problem with your approach is that in 7.21 Silabs changed the location of a fire fields a bit and this can break the dongle if you write pre-7.21 NVM. We even do the NVM conversion when the dongle is upgraded to 7.21 (not during restore - just during a normal firmware upgrade). Because nobody can guarantee that the format remains same. Our approach keeps things separated: fixed API on the Serial API and the dongle is mapping it inside the firmware to the right firmware version specific place.

**Reply by Botched1** [reply]
> Maybe on Friday.

No problem. I got my existing system working, so the timing is less urgent. I'll be happy to test whenever it is ready - no rush.

**Reply by PoltoS** [reply]
Forgot to mention, this is now public: https://z-wave.me/support/controller-tools/
There is also a newer version in https://z-wave.me/support/controller-tools/beta
@Botched1 @AlCalzone

**Reply by Botched1** [reply]
Thanks. I used it to upgrade firmware from 7.40 to 7.42, and then 7.42 to 7.45. No issues.

I'll try the migration/porting function soon.

EDIT: Sigh, never mind. no Long Range support in the migration/porting, and >33% of my network is Long Range. Will that be added later @PoltoS ? I would love to use my z-station more, but simply can not until I can migrate my 100+ node network to it successfully.
![image](https://github.com/user-attachments/assets/2fd8fccc-78f3-4f22-88e2-371d2299fded)

**Reply by PoltoS** [reply]
You have 30 LR devices? Hm.. I've never tried of the secondary gateway gets LR devices in the list of existing devices. Probably not....

But if you are on Z-Wave JS, you can construct the backup and wrote directly to the memory. I think, @AlCalzone wanted to implement it. Anyway, this will require some time to implement. This was made to let people migrate from old 5th gen networks and discontinued controllers to new 7th/8th. Never trie d to migrate an LR network

**Reply by Botched1** [reply]
> You have 30 LR devices?

I was a bit off. 26 right now, but 10 more I haven't paired/replaced the mesh device on yet.

> I've never tried of the secondary gateway gets LR devices in the list of existing devices. Probably not....

It does not.

> Anyway, this will require some time to implement. This was made to let people migrate from old 5th gen networks and discontinued controllers to new 7th/8th. Never tried to migrate an LR network

No problem. Z-Station has to go back to the box for now. :) I simply do not have time to re-pair 100+ devices.

**Reply by Botched1** [reply]
Side note @AlCalzone .... I have no idea what form factor the upcoming Home Assistant z-wave hardware will take, but if it followed the Z-Station form factor, but with non-proprietary firmware, I would buy it in a SECOND.

z-wave.me has the perfect hardware form factor, but the proprietary firmware / proprietary tools kills it for many.

**Reply by AlCalzone** [reply]
I'll have to take a look at the Z-Wave.me NVMs, because normally Z-Wave JS should be able to restore the NVM backups including LR nodes, unless there's something that differs from the NVM formats the official firmwares use.

> I have no idea what form factor the upcoming Home Assistant z-wave hardware will take

Cannot say for sure, but likely quite a bit bigger.

**Reply by Botched1** [reply]
> Cannot say for sure, but likely quite a bit bigger.

Interesting. As long as they are not USB dongle format and preferably with external antenna, I would be happy. :)

**Reply by Botched1** [reply]
It is frustrating @PoltoS ... The Z-Station is probably the nicest / best hardware form factor I have seen on a z-wave device, but the custom bootloader and firmware prevents many of us from using it as we can't migrate to it (if using LR, which many US people do now).

Would you ever consider releasing a standard/non-proprietary bootloader and firmware for the Z-Station so that the standard PC Tools / ZWave JS tools could be used natively?
